src/repositories/shopping-cart.repository.ts

```diff
@@ -4,14 +4,36 @@
 // License text available at https://opensource.org/licenses/MIT
 
 import {DefaultKeyValueRepository} from '@loopback/repository';
-import {ShoppingCart} from '../models/shopping-cart.model';
+import {ShoppingCart, ShoppingCartItem} from '../models/shopping-cart.model';
 import {RedisDataSource} from '../datasources/redis.datasource';
 import {inject} from '@loopback/context';
+import {promisify} from 'util';
 
 export class ShoppingCartRepository extends DefaultKeyValueRepository<
   ShoppingCart
 > {
   constructor(@inject('datasources.redis') ds: RedisDataSource) {
     super(ShoppingCart, ds);
   }
+
+  /**
+   * Add an item to the shopping cart with optimistic lock
+   * @param userId User id
+   * @param item Item to be added
+   */
+  async addItem(userId: string, item: ShoppingCartItem) {
+    const connector = this.kvModelClass.dataSource!.connector!;
+    // tslint:disable-next-line:no-any
+    const execute = promisify((cmd: string, args: any[], cb: Function) => {
```

Comment:
I'm not sure I follow what `execute` does? Apart for this and it's usage, the rest of it looks good.

Reply:
But we're still using `this.get`, `this.set`? So what do the commands do?

Reply:
this.get/this.set delegate to execute behind the scene.

Reply:
The reason to use `execute` is that we don't have a `this.checkAndSet`. Maybe we should add it to the KVRepository, such as:
```ts
checkAndSet(key, current => {...});
```

Reply:
I'm considering the multiple device case. A user can access the app from multiple devices that share the same cart.

Reply:
But each call does a `get` and then `set`. Is a user really going to generate multiple requests at the same time from multiple devices? 

Either way, I'm ok with the change -- can you just add a comment stating the reason for using a transaction so it's clear to anyone looking at this example.

Reply:
Added more comments.